packages/provider/src/MultiChainProvider.test.ts

```
@@ -0,0 +1,404 @@
+          method: 'metamask_handshake',
+          params: {
+            requiredNamespaces: {
+              eip155: {
+                chains: ['eip155:1'],
+                methods: [],
+                events: [],
+              },
+            },
+          },
+        },
+      });
+    });
+
+    it('verifies the JSON-RPC response', async () => {
+      const request = ethereum.request as jest.MockedFunction<
+        typeof ethereum.request
+      >;
+
+      request.mockImplementation(async () => ({
+        jsonrpc: '2.0',
+        id: 1,
+        result: {
+          namespaces: {
+            eip155: 'foo',
+          },
+        },
+      }));
+
+      const provider = new MultiChainProvider();
+      expect(provider.isConnected).toBe(false);
+
+      const listener = jest.fn();
+      provider.on('session_update', listener);
+
+      const { approval } = await provider.connect({ requiredNamespaces: {} });
+      await expect(approval()).rejects.toThrow(
+        'Invalid session: At path: namespaces.eip155 -- Expected an object, but received: "foo"',
+      );
+
+      expect(provider.isConnected).toBe(false);
+      expect(listener).not.toHaveBeenCalled();
+    });
+
+    it('throws on JSON-RPC error response', async () => {
+      const request = ethereum.request as jest.MockedFunction<
+        typeof ethereum.request
+      >;
+
+      request.mockImplementation(async () => ({
+        jsonrpc: '2.0',
+        id: 1,
+        error: {
+          code: -1,
+          message: 'foo',
+        },
+      }));
+
+      const provider = new MultiChainProvider();
+      expect(provider.isConnected).toBe(false);
+
+      const { approval } = await provider.connect({ requiredNamespaces: {} });
+      await expect(approval()).rejects.toThrow(
+        'Not a successful JSON-RPC response: At path: result -- Expected the value to satisfy a union of `literal | boolean | number | string | lazy | lazy`, but received: undefined.',
```

Could we throw the request error instead? That would be more descriptive